### PR TITLE
Optimize asset table rendering

### PR DESCRIPTION
**Pull request by somebody1234**
### Pull Request Description
- Depends on:
  - #11380
- Some optimizations for re-rendering assets table:
  - Omit `visibilities` from `state` in favor of passing each `AssetRow`'s `visibility` directly to the row. This minimizes spurious `state` updates.
  - Pass `id` and `parentId` to `AssetRow` instead of the entire object. This ensures that re-fetches do not force a re-render of the `AssetRow` - we are no longer passing a reference to the object, so we are now comparing by string comparison (which is almost always stable).

### Important Notes
None

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] ~~The documentation has been updated, if necessary.~~
- [x] ~~Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.~~
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] ~~Unit tests have been written where possible.~~
- [x] ~~If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.~~

### CODE REVIEW

**Comment by somebody1234** [comment]
@MrFlashAccount for code review

File: app/common/src/utilities/data/object.ts [hunk 1]
```diff
@@ -9,6 +9,26 @@ export type Mutable<T> = {
   -readonly [K in keyof T]: T[K]
 }
 
+/** Whether two values are deeply equal. */
+export function deepEqual<T>(a: T, b: T) {
+  if (Object.is(a, b)) return true
+  if (typeof a !== typeof b) return false
+  if (typeof a !== 'object' || typeof b !== 'object') return Object.is(a, b)
```
Comment (radeusgd) [inline app/common/src/utilities/data/object.ts]:
```suggestion
  if (Object.is(a, b)) return true
  if (typeof a !== typeof b) return false
  if (typeof a !== 'object' || typeof b !== 'object') return false
```
once we get to the 3rd `if`, we know that `Object.is(a, b) == false`. If it were `true`, we'd have returned from the first `if`.

File: app/common/src/utilities/data/object.ts [hunk 2]
```diff
@@ -9,6 +9,26 @@ export type Mutable<T> = {
...
+  if (Array.isArray(a) && Array.isArray(b) && a.length !== b.length) return false
+  // They cannot both be null because the `Object.is(a, b)` case above ensures they are not equal.
+  if (a == null || b == null) return false
+  for (const k in a) {
+    if (!(k in b)) return false
+  }
+  for (const k in b) {
+    if (!(k in a)) return true
+  }
```
Comment (radeusgd) [inline app/common/src/utilities/data/object.ts]: This asymmetry looks suspicious

File: app/common/src/utilities/data/object.ts [hunk 3]
```diff
@@ -9,6 +9,26 @@ export type Mutable<T> = {
...
+  for (const k in a) {
+    if (!deepEqual(a[k], b[k])) return false
+  }
+  return false
```
Comment (radeusgd) [inline app/common/src/utilities/data/object.ts]: shouldn't this return `true` actually?

Comment (somebody1234) [reply app/common/src/utilities/data/object.ts]: yeah i uh. have no idea how these bugs happened. tired i guess...

File: app/gui/src/dashboard/hooks/backendHooks.ts [hunk 4]
```diff
@@ -39,6 +39,7 @@ import {
 import { TEAMS_DIRECTORY_ID, USERS_DIRECTORY_ID } from '#/services/remoteBackendPaths'
 import { usePreventNavigation } from '#/utilities/preventNavigation'
 import { toRfc3339 } from 'enso-common/src/utilities/data/dateTime'
+import { deepEqual } from 'enso-common/src/utilities/data/object'
```
Comment (MrFlashAccount) [inline app/gui/src/dashboard/hooks/backendHooks.ts]: I doubt we need to use deepEqual here. I did some measures over react-compiler and it actually slows down rendering comparing to just returning the result

Comment (somebody1234) [reply app/gui/src/dashboard/hooks/backendHooks.ts]: removed memoization (?), i guess it'll be good if you could measure again and make sure performance is still fine

Comment (MrFlashAccount) [reply app/gui/src/dashboard/hooks/backendHooks.ts]: I did the measurement on the local branch, now much better!

**Review by MrFlashAccount — APPROVED** [review]

**Comment by MrFlashAccount** [comment]
@somebody1234 I think next steps might be:
1. Put the `state` prop into `zustand` state, add fine-grained access to the props (it's save to put the `AssetTree` there as long as we don't mutate it)
2. Generate `flat` structure of the `AssetTree` during creation, and not afterwards (I found that `preorderTraversal` takes significant amount of time on large tree structures

**Comment by somebody1234** [comment]
2. Generate flat structure of the `AssetTree` during creation

not sure i agree. i think it's about time to remove `AssetTreeNode` completely and just rely on recursive components, each with its own query. what do you think?


also unrelated, thoughts about maybe doing more fine-grained `listDirectory` and `listProjectSessions` (etc) invalidation eventually?

**Comment by MrFlashAccount** [comment]
2. On my experience, having a flat tree structure makes it easier to add virtualization (since the parent component might be offscreen and thus unmounted with all the nested components). Having a flat structure makes things easier. This could possibly be avoided by leveraging `sticky` functionality, but it makes things more complicated than they should be, see: https://tanstack.com/virtual/latest/docs/framework/react/examples/sticky

> also unrelated, thoughts about maybe doing more fine-grained listDirectory and listProjectSessions (etc) invalidation eventually?

Agree, though this will only improve the network side, but not the rendering itself

**Comment by MrFlashAccount** [comment]
IMO, it makes sence to have a tree structure (makes it easier to track parent -> child relations), but it shall be _completely_ immutable

**Comment by somebody1234** [comment]
more importantly with 2. - that means that sorting will no longer work, right?

**Comment by MrFlashAccount** [comment]
> more importantly with 2. - that means that sorting will no longer work, right?

Why?

**Comment by somebody1234** [comment]
we want to sort assets within directories - will having a flat structure still let us do that?

**Comment by MrFlashAccount** [comment]
> we want to sort assets within directories - will having a flat structure still let us do that?

Since we're not changing the current behavior (we'll still have the both `tree`(`assetTree` - https://github.com/enso-org/enso/blob/c87d130bcf80d36869c65454f3ca01c91128c939/app/gui/src/dashboard/layouts/AssetsTable.tsx#L520) and `flat`(`displayItems`- https://github.com/enso-org/enso/blob/c87d130bcf80d36869c65454f3ca01c91128c939/app/gui/src/dashboard/layouts/AssetsTable.tsx#L793) structures  the sorting functionality remains unchanged.

My point is to join 2 steps (build the tree and generate flat structure based on the tree) into a single one - on generating the tree. Ideally I'd like to also to put things into the right place based on the sorting preferences)

Ideal time complexity is O(n), but O(nlogn) is acceptable at this point.

**Comment by hubertp** [comment]
@radeusgd has experienced occasional flickering (showing up/disappearing) of cloud projects. Will this PR address this by any chance?

**Comment by somebody1234** [comment]
@hubertp there is a follow up PR:

- #11405

that should improve performance much more on top of this one - depending on what is meant by flickering then that one may help

File: app/common/src/services/Backend.ts [hunk 5]
```diff
@@ -823,11 +818,19 @@ export function createRootDirectoryAsset(directoryId: DirectoryId): DirectoryAss
...
+    parentsPath: '',
+    virtualParentsPath: '',
   }
 }
 
+/** Extract the file extension from a file name. */
+function fileExtension(fileNameOrPath: string) {
+  return fileNameOrPath.match(/[.]([^.]+?)$/)?.[1] ?? ''
+}
+
```
Comment (farmaazon) [inline app/common/src/services/Backend.ts:833]: There's also [path.extname](https://www.w3schools.com/nodejs/met_path_extname.asp), however I'm not sure if available here.

File: app/common/src/utilities/data/object.ts [hunk 6]
```diff
@@ -9,6 +9,28 @@ export type Mutable<T> = {
   -readonly [K in keyof T]: T[K]
 }
 
+/** Whether two values are deeply equal. */
```
Comment (farmaazon) [inline app/common/src/utilities/data/object.ts]: I think the docs should be extended, as "deeply equal" is unprecise in js world.

This very instance, for example, will consider two objects of different classes equal if they happen to have same properties.

Also, there's already an existing method in [lodash library](https://lodash.com/docs/4.17.15#isEqual) which AFAIK we have an indirect dependency to anyway.

**Review by farmaazon — APPROVED** [review]

File: app/common/src/services/Backend.ts (hunk 5 above)
Comment (somebody1234) [reply app/common/src/services/Backend.ts:833]: indeed not. we could theoretically polyfill node's builtin modules, but i guess i'd rather not

File: app/common/src/utilities/data/object.ts (hunk 6 above)
Comment (somebody1234) [reply app/common/src/utilities/data/object.ts]: added [fast-deep-equal](https://www.npmjs.com/package/fast-deep-equal) instead. note that it's no longer actually used currently but i don't think it hurts to add this anyway for posterity. thoughts?

Comment (farmaazon) [reply app/common/src/utilities/data/object.ts]: I don't see much point, unless you actually see that it will be needed soon.

Comment (somebody1234) [reply app/common/src/utilities/data/object.ts]: yeah fair enough. i don't actually see it being needed

File: app/gui/src/dashboard/layouts/AssetsTable/EventListProvider.tsx [hunk 7]
```diff
@@ -170,6 +171,7 @@ export function useAssetListEventListener(
     const events = initialEventsRef.current
     if (events && !alreadyRun) {
       // Event handlers are not idempotent and MUST NOT be handled twice.
+      // eslint-disable-next-line react-compiler/react-compiler
```
Comment (MrFlashAccount) [inline app/gui/src/dashboard/layouts/AssetsTable/EventListProvider.tsx:174]: Not sure we should suppress the react-compiler in that case.
The component is huge, and one line shouldn't lead to deopt for the component. IMO, better either leave the warning or refactor, e.g. fix or extract the deopt code into a separate hook/component

File: app/gui/src/dashboard/layouts/AssetsTable.tsx [hunk 8]
```diff
@@ -1192,6 +1192,7 @@ export default function AssetsTable(props: AssetsTableProps) {
       }
     }
     // This effect MUST only run when `initialProjectName` is changed.
+    // eslint-disable-next-line react-compiler/react-compiler
```
Comment (MrFlashAccount) [inline app/gui/src/dashboard/layouts/AssetsTable.tsx]: Not sure we should suppress the react-compiler here. AssetsTable is a huge component, and one line shouldn't deopt the whole component.
IMO, there are 2 ways - fix the issue or extract the deopt code into a separate component/hook.

Same goes for the other cases

Comment (somebody1234) [reply app/gui/src/dashboard/layouts/AssetsTable.tsx]: ah you do have a point, forgot about that...

**Comment by somebody1234** [comment]
@MrFlashAccount react compiler fixes should be ready for re-review?

**Comment by somebody1234** [comment]
some hacky workarounds which i'm not 100% sure about...

**Comment by MrFlashAccount** [comment]
I’m a bit worried that the fixes for react-compiler shall be tested separately. But this PR is an essential part of the compiler. Can we extract the fixes for react-compiler lint and merge them after landing these 2 PRs?

**Comment by MrFlashAccount** [comment]
Let's do the QA and merge ✨

**Pull request merged**